### PR TITLE
NEWRELIC-4431 Enhance cassandra-driver to include instance metadata

### PR DESCRIPTION
**Pull request by jmartin4563**
## Proposed Release Notes
* Fixed `cassandra-driver` instrumentation to properly set instance details on query segments/spans.

## Links
Closes NEWRELIC-4431

## Details
* Added necessary instrumentation to ensure capture of metadata. `RequestExecution.start` is instrumented in older versions to ensure persisting the correct context down the chain. `RequestExecution. _sendOnConnection` is instrumented in all supported versions as it has the information we need and is used by all the query methods.
* Updated instrumentation to be more explicit about version support and instrumentation differences
* Added versioned tests asserting that metadata is applied to the appropriate span
* Added JSDocs, since we talked about cleaning up older code as we go in our retro

### CODE REVIEW

**Review by bizob2828 — CHANGES_REQUESTED** [review]
Nice work digging into all the shim magic to figure this one out.  A few comments but overall looking good!

File: lib/instrumentation/cassandra-driver.js [hunk 1]
```diff
@@ -5,16 +5,82 @@
...
+      query: shim.FIRST,
+      callback: shim.LAST
+    })
+
+    shim.wrap(RequestExecutionProto, 'start', function wrapStart(shim, start) {
+      return function wrappedStart() {
+        const parent = shim.getSegment()
+        const self = this
+
+        arguments[0] = shim.wrap(
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/instrumentation/cassandra-driver.js:61]: The reason for this is arguments isn't actually an array but an object. it'd still work but this is just a safe thing we do

File: test/versioned/cassandra-driver/query.tap.js [hunk 2]
```diff
@@ -8,6 +8,7 @@
 const test = require('tap').test
 const params = require('../../lib/params')
 const helper = require('../../lib/agent_helper')
+const os = require('os')
 
 const agent = helper.instrumentMockedAgent()
 const cassandra = require('cassandra-driver')
```
[the remark this answers is not in the document]
Comment (jmartin4563) [reply test/versioned/cassandra-driver/query.tap.js:14]: 👍🏻 agreed

[the remark this answers is not in the document]
Comment (jmartin4563) [reply test/versioned/cassandra-driver/query.tap.js:14]: https://issues.newrelic.com/browse/NEWRELIC-4671

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**